Source/Core/DolphinQt2/FIFOWindow.cpp

```diff
@@ -0,0 +1,359 @@
+          &FIFOWindow::OnFrameFromChanged);
+  connect(m_frame_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnFrameToChanged);
+
+  connect(m_object_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnObjectFromChanged);
+  connect(m_object_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnObjectToChanged);
+}
+
+void FIFOWindow::LoadRecording()
+{
+  QString path = QFileDialog::getOpenFileName(this, tr("Open FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!m_recording)
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(), [] {
+      QueueOnObject(&Settings::Instance(), [] { Settings::Instance().TriggerFIFORecordingDone(); });
+    });
+
+    m_save->setEnabled(false);
+    m_recording = true;
+    m_recording_done = false;
+
+    m_record->setText(tr("Stop"));
+  }
+  else
+  {
+    recorder.StopRecording();
+  }
+
+  UpdateInfo();
+}
+
+void FIFOWindow::OnEmulationStarted()
+{
+  m_record->setEnabled(!m_playing);
+  m_load->setEnabled(false);
+}
+
+void FIFOWindow::OnEmulationStopped()
+{
+  m_record->setEnabled(false);
+  m_load->setEnabled(true);
+
+  // If we have previously been recording, stop now.
+  if (m_recording)
+    ToggleRecording();
+
+  if (m_playing)
+  {
+    m_record->setEnabled(true);
+    m_playing = false;
+  }
+}
+
+void FIFOWindow::OnRecordingDone()
+{
+  m_recording = false;
+  m_recording_done = true;
+
+  UpdateInfo();
+
+  m_save->setEnabled(true);
+  m_record->setText(tr("Record"));
+}
+
+void FIFOWindow::UpdateInfo()
+{
+  if (m_playing)
+  {
+    FifoDataFile* file = FifoPlayer::GetInstance().GetFile();
+    m_info_label->setText(
+        tr("%1 frame(s)\n%2 object(s)\nCurrent Frame: %3")
+            .arg(QString::number(file->GetFrameCount()),
+                 QString::number(FifoPlayer::GetInstance().GetFrameObjectCount()),
+                 QString::number(FifoPlayer::GetInstance().GetCurrentFrameNum())));
+    return;
+  }
+
+  if (m_recording)
+  {
+    m_info_label->setText(tr("Recording..."));
+    return;
+  }
+
+  if (m_recording_done)
+  {
+    FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
```

Comment:
Does this logic need to be run on every call to `UpdateInfo`? It looks like it could be moved into `OnRecordingDone()`.